plugins/utils/todo-comments.nix

```diff
@@ -16,275 +24,394 @@ let
+            key = mkOption {
+              type = types.str;
+              description = "Key for the `${funcName}` function.";
             };
-        in
-        mapAttrs mkKeymapOption commands;
+
+            cwd = mkOption {
+              type = types.nullOr types.str;
+              description = "Specify the directory to search for comments";
+              default = null;
+              example = "~/projects/foobar";
+            };
+
+            keywords = mkOption {
+              type = types.nullOr types.str;
+              description = ''
+                Comma separated list of keywords to filter results by.
+                Keywords are case-sensitive.
+              '';
+              default = null;
+              example = "TODO,FIX";
+            };
+
+            options = keymaps.mapConfigOptions;
+          };
+      in
+      mapAttrs mkKeymapOption commands;
+
+    ripgrepPackage = mkPackageOption {
+      name = "ripgrep";
+      default = pkgs.ripgrep;
     };
   };
 
-  config =
-    let
-      setupOptions = {
-        inherit (cfg) signs;
-        sign_priority = cfg.signPriority;
-        inherit (cfg) keywords;
-        gui_style = cfg.guiStyle;
-        merge_keywords = cfg.mergeKeywords;
-        highlight = {
-          inherit (cfg.highlight)
-            multiline
-            before
-            keyword
-            after
-            exclude
-            ;
-          pattern = helpers.ifNonNull' cfg.highlight.pattern (helpers.mkRaw "[[${cfg.highlight.pattern}]]");
-          multiline_pattern = cfg.highlight.multilinePattern;
-          multiline_context = cfg.highlight.multilineContext;
-          comments_only = cfg.highlight.commentsOnly;
-          max_line_len = cfg.highlight.maxLineLen;
-        };
-        inherit (cfg) colors;
-        search = {
-          inherit (cfg.search) command args;
-          pattern = helpers.ifNonNull' cfg.search.pattern (
-            if isList cfg.search.pattern then
-              (map helpers.mkRaw cfg.search.pattern)
-            else
-              helpers.mkRaw "[[${cfg.search.pattern}]]"
-          );
-        };
-      } // cfg.extraOptions;
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-      extraPackages = [ cfg.ripgrepPackage ];
-      extraConfigLua = ''
-        require("todo-comments").setup${helpers.toLuaObject setupOptions}
-      '';
+  extraConfig = cfg: {
+    extraPackages = [ cfg.ripgrepPackage ];
+
+    keymaps = flatten (
+      mapAttrsToList (
+        optionName: funcName:
+        let
+          keymap = cfg.keymaps.${optionName};
+
+          cwd = optionalString (keymap.cwd != null) " cwd=${keymap.cwd}";
+          keywords = optionalString (keymap.keywords != null) " keywords=${keymap.keywords}";
+        in
+        optional (keymap != null) {
+          mode = "n";
+          inherit (keymap) key options;
+          action = ":${funcName}${cwd}${keywords}<CR>";
+        }
+      ) commands
+    );
```

Comment:
I see this is unchanged from the previous implementation, other than being indented further.

However, I think this would be simpler mapping over `cfg.keymaps` instead of mapping over `commands`?

Ideally, I'd like to avoid using `flatten`, perhaps we could do something like:

```nix
keymaps = lib.pipe cfg.keymaps [
  (filterAttrs (n: v: v != null))
  (mapAttrsToList (name: keymap: {
    inherit (keymap) key option;
    mode = "n";
    action =
      let
        cmd = commands.${name};
        cwd = optionalString (keymap.cwd != null) " cwd=${keymap.cwd}";
        keywords = optionalString (keymap.keywords != null) " keywords=${keymap.keywords}";
      in
      "<cmd>${cmd}${cwd}${keywords}<cr>";
  }))
];
```

Looking at the `commands` attrset, we could probably replace it with a `upperFirstChar` style helper.

However, many of the details would be slightly different if we can go down the #603 route.

Reply:
> I see this is unchanged from the previous implementation, other than being indented further.
> 
Yeah, sorry I didn't touch this bit. I wasn't sure what to do with it, yet.

I think the suggested solution is a lot cleaner.  But, I need to get more familiar with that keymap issue that was mentioned here and above.

Reply:
We can always tackle it in another PR, but the other suggestions above could be better.

Reply:
If I can tackle that separately it would give me a bit to figure it out and maybe get the extraOptions support for mkMapOptionSubmodule